parser-typechecker/src/Unison/PatternMatchCoverage/UFMap.hs

```diff
@@ -0,0 +1,234 @@
+  -- * the @UFMap@ after path compression
+  Maybe (k, Int, v, UFMap k v)
+lookupCanon k m =
+  getCompose (alterF k nothing just m)
+  where
+    nothing = Compose Nothing
+    just k s v = Compose (Just (k, s, v, Canonical s v))
+
+data UnionHaltReason k v
+  = KeyNotFound k
+  | MergeFailed v v
+
+data UnionValue k v a
+  = UnionValue k Int v (UFValue k v) a
+  deriving stock (Functor)
+
+union ::
+  forall m k v r.
+  (MonadFix m, Ord k) =>
+  k ->
+  k ->
+  UFMap k v ->
+  (k -> v -> UFMap k v -> m (Maybe r)) ->
+  m (Maybe r)
+union k0 k1 mapinit mergeValues = toMaybe do
+  rec let lu ::
+            k ->
+            UFMap k v ->
+            (k -> UFMap k v -> Maybe (Compose (Either (UnionHaltReason k v)) (UnionValue k v) (UFMap k v))) ->
+            Either (UnionHaltReason k v) (UnionValue k v (UFMap k v))
+          lu k m loopGuard = getCompose (alterFWithHalt k loopGuard luNothing luJust m)
+            where
+              luNothing = Compose (Left (KeyNotFound k))
+          luJust k s v =
+            -- a final value thunk is inserted before it is resolved,
+            -- as the final result cannot be known before we have
+            -- looked up both values and merged them
+            let newValue =
+                  let newSize = case kcanon0 == kcanon1 of
+                        True -> size0
+                        False -> size0 + size1
+                   in case chosenCanon == k of
+                        True -> Canonical newSize canonValue
+                        False -> Indirection chosenCanon
+             in Compose (Right (UnionValue k s v newValue newValue))
+      UnionValue kcanon0 size0 v0 vfinal0 map0 <- ExceptT $ pure $ lu k0 mapinit \_ _ -> Nothing
+      UnionValue kcanon1 size1 v1 vfinal1 map1 <- ExceptT $
+        pure $ lu k1 map0 \k m -> case k == kcanon0 of
+          False -> Nothing
+          True -> Just (Compose (Right (UnionValue k size0 v0 vfinal0 m)))
+      -- Join the smaller equivalence class to the larger to bound
+      -- worst case number of lookups to log(n). This is the same
+      -- strategy as the weighted fast-union algorithm.
+      let (chosenCanon, canonValue, nonCanonValue) = case size0 > size1 of
+            True -> (kcanon0, v0, v1)
+            False -> (kcanon1, v1, v0)
+  map2 <-
+    let res =
+          ExceptT $
+            mergeValues chosenCanon nonCanonValue map1 <&> \case
+              Nothing -> Left (MergeFailed v0 v1)
+              Just x -> Right x
+     in -- Now that both lookups have completed we can safely force the
+        -- final values
+        vfinal0 `seq` vfinal1 `seq` res
+  pure map2
+  where
+    toMaybe :: ExceptT (UnionHaltReason k v) m r -> m (Maybe r)
+    toMaybe (ExceptT action) =
+      action <&> \case
+        Right m -> Just m
+        Left r -> case r of
+          KeyNotFound _k -> Nothing
+          MergeFailed _v0 _v1 -> Nothing
+
+toClasses :: forall k v. (Ord k) => UFMap k v -> [(k, Set k, v)]
+toClasses m0 =
```

Comment:
I know some of these names probably come straight out of the paper, but if I ever have to patch something in here I'm definitely not going to know what the intended purpose of `phi` is haha.

Would you mind expanding some of the binding names and add a couple comments, type sigs and maybe doc-string?